File: Delphi/NewsGroups/MessageForm.pas
unit MessageForm;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, clMailMessage, StdCtrls;

type
  TclMessageForm = class(TForm)
    Label1: TLabel;
    Label2: TLabel;
    Label3: TLabel;
    edtFrom: TEdit;
    edtSubject: TEdit;
    memText: TMemo;
    btnPost: TButton;
    btnCancel: TButton;
  private
    { Private declarations }
  public
    class function PostNew(Article: TclMailMessage): Boolean;
    class function ReplyTo(Article: TclMailMessage): Boolean;
  end;

implementation

{$R *.dfm}

{ TclMessageForm }

class function TclMessageForm.PostNew(Article: TclMailMessage): Boolean;
var
  dlg: TclMessageForm;
begin
  dlg := TclMessageForm.Create(nil);
  try
    Result := dlg.ShowModal() = mrOK;
    if Result then
    begin
      Article.BuildMessage(dlg.memText.Lines.Text, '');
      Article.Subject := dlg.edtSubject.Text;
      Article.From.FullAddress := dlg.edtFrom.Text;
    end;
  finally
    dlg.Free();
  end;
end;

class function TclMessageForm.ReplyTo(Article: TclMailMessage): Boolean;
var
  i: Integer;
  src: TStrings;
  refs: TStrings;
  dlg: TclMessageForm;
begin
  dlg := TclMessageForm.Create(nil);
  src := TStringList.Create();
  refs := TStringList.Create();
  try
    src.Assign(Article.MessageText);
    for i := 0 to src.Count - 1 do
    begin
      src[i] := '> ' + src[i];
    end;
    src.Insert(0, '');
    src.Insert(0, '');
    dlg.memText.Lines.Assign(src);

    if (Pos('Re:', Article.Subject) <> 1) then
    begin
      dlg.edtSubject.Text := 'Re: ' + Article.Subject;
    end;

    refs.Assign(Article.References);

    dlg.btnPost.Caption := 'Reply';

    Result := dlg.ShowModal() = mrOK;
    if Result then
    begin
      Article.BuildMessage(dlg.memText.Lines.Text, '');
      Article.Subject := dlg.edtSubject.Text;
      Article.From.FullAddress := dlg.edtFrom.Text;
      Article.References.Assign(refs);
    end;
  finally
    refs.Free();
    src.Free();
    dlg.Free();
  end;
end;

end.

File: CBuilder/NewsGroups/MessageForm.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "MessageForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
//---------------------------------------------------------------------------
__fastcall TclMessageForm::TclMessageForm(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------

bool __fastcall TclMessageForm::PostNew(TclMailMessage *Article)
{
  bool res;
  TclMessageForm *dlg = new TclMessageForm(NULL);
  __try
  {
    res = (dlg->ShowModal() == mrOk);
    if(res)
    {
      Article->BuildMessage(dlg->memText->Lines->Text, "");
      Article->Subject = dlg->edtSubject->Text;
      Article->From->FullAddress = dlg->edtFrom->Text;
    }
  }
  __finally
  {
    delete dlg;
  }
  return res;
}
//---------------------------------------------------------------------------

bool __fastcall TclMessageForm::ReplyTo(TclMailMessage *Article)
{
  bool res;
  TclMessageForm *dlg = new TclMessageForm(NULL);
  TStrings *src = new TStringList();
  TStrings *refs = new TStringList();
  __try
  {
    src->Assign(Article->MessageText);
    for(int i = 0; i < src->Count; i++)
      src->Strings[i] = "> " + src->Strings[i];
    src->Insert(0, "");
    src->Insert(0, "");
    dlg->memText->Lines->Assign(src);

    if(Article->Subject.Pos("Re:") != 1)
      dlg->edtSubject->Text = "Re: " + Article->Subject;

    refs->Assign(Article->References);

    dlg->btnPost->Caption = "Reply";

    res = (dlg->ShowModal() == mrOk);
    if(res)
    {
      Article->BuildMessage(dlg->memText->Lines->Text, "");
      Article->Subject = dlg->edtSubject->Text;
      Article->From->FullAddress = dlg->edtFrom->Text;
      Article->References->Assign(refs);
    }
  }
  __finally
  {
    delete refs;
    delete src;
    delete dlg;
  }
  return res;
}
//---------------------------------------------------------------------------

